WIP: total redo of RabbitMqIO for Beam

drobert opened the pull request:
## Why

We've had no end of trouble with the RabbitIO connector for Beam. CheckpointMark's cannot safely advance and be canceled, correlation ids are required but misused as a deduping mechanism, watermarks do not consistently advance, etc.

For some background reading, see this thread on the Beam Dev DL: https://lists.apache.org/thread.html/7f8f7e3784816679da2120c1c6c0163e983c4f6a3e5339d1a5b624f7@%3Cdev.beam.apache.org%3E

## What

The goal is to make some foundational changes to the RabbitMqIO connector for Beam, to accomplish at a minimum the following:
- do not lose messages (CheckpointMark is currently unusable as implemented as it cannot be serialized and resumed properly).
- allow for custom event times, separate from processing time
- model connection management after JdbcIo to use a single Connection per host, and multiplex over channels by 'client' (for both Read and Write)
  - using a 'loan pattern' for channels for safe management
- using timestamp/watermark policies modeled after KafkaIOs
- ripped out a lot of classes from RabbitMqIO into separate files, as KafkaIO does, but also because it's just hard to navigate
- reworked the queue/exchange paradigm significantly, based on a unified `ReadParadigm` concept. I also eliminated a few previously-supported use cases because they're not really a good fit for Beam (ephemeral/auto-delete queues, declaring exchanges, and a few other things).
- uses the amqp pull-based api (`channel.basicGet`) which also allows for a server-supplied estimate of current queue depth, used for watermark calculations

## Notes for Reviewers

- I would start with either `UnboundedRabbitMqReader` or `RabbitMqCheckpointMark` to see the notable changes. Of these, the important aspects are:
  - Connection/Channel lifecycle management
  - TimestampPolicy and watermark calculations
- I'd look at the timestamp policy and record id policies next. The timestamp policy was largely lifted from Kafka's. See https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/CustomTimestampPolicyWithLimitedDelay.java

I've tried to add grokable javadoc to everything I've touched. If things don't make sense, please say something.

## TODO

- [x] I *just noticed* AMQP's `GetResponse` has a `messageCount` property defined as "the server's most recent estimate of the number of messages remaining on the queue". This would allow us to implement timestamp/watermark almost identically to how kafka does it. Very exciting.
- [x] classes (non-test) should compile
- [x] Implement pluggable strategy for deduping messages
  - [ ] possible todo: fallback strategies (try message id, barring that try correlation id, barring that, etc.)
- [x] Implement pluggable strategy for extracting timestamps and setting watermarks (base on the [KafkaIO TimestampPolicyFactory abstraction](https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/TimestampPolicyFactory.java)
- [x] Wire up the pluggable strategies to the UnboundedSourceReader
- [ ] Update all the RabbitMqIO javadoc/documentation
- [x] Handle the impossible serialization/deserialization of a checkpoint mark (this will rely on an appropriate 'dedupe' strategy and pluggable watermark advancements)
- [x] eliminate the notion of 'queue vs exchange'; this is totally wrong
- [x] so. much. testing.

Commit:
Remove amqp tx. No purpose.

Commit:
Adding RecordIdPolicy and TimestampPolicy concepts

Commit:
Dynamic record id and timestamp policies.

Timestamp policies modeled after KafkaIO's.
Record Id policies flexible, appropriate for rabbit.

TODO:
- connection management
- serialize/resume of a CheckpointMark
- update documentation
- so. much. testing.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -123,9 +124,11 @@ public static Read read() {
         .setQueueDeclare(false)
         .setExchangeDeclare(false)
+        // this is not a great policy, this should be set to something more sensible
+        .setRecordIdPolicy(RecordIdPolicy.alwaysUnique())
```
I'm actually tempted to get rid of this policy and default to one of the following:
1. sha256
2. always fail immediately and loudly

Deduping messages is really mandatory for rabbit, since CheckpointMark will often fail to serialize state with unacked messages and they *will* be reprocessed.

NJAldwin replied:
😬 hm, I'm not sure exactly what is best here.  Any thoughts on this from the Beam folks as far as what Beam consumers might expect?  I'd lean towards always failing with an actionable message, personally.

drobert replied:
I'll ask. I'm not sure what I think is appropriate. I think the best thing to do would be default to a chain-of-command with this priority:
- message id
- timestamp + sha256 (if message id property not set)
- explode loudly (if message id and timestamp properties not set)

drobert replied:
Current plan is to address this via documentation (which I've written in class-level documentation).

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -409,10 +440,11 @@ public Read withUseCorrelationId(boolean useCorrelationId) {
     @Override
     public boolean requiresDeduping() {
-      return spec.useCorrelationId();
+      return true;
```
This *cannot* be turned off.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -436,18 +468,40 @@ public void finalizeCheckpoint() throws IOException {
     }
   }
 
+  private static class TimestampPolicyContext extends TimestampPolicy.LastRead {
```
Some static factory functions would make this easier to use. Will add.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -546,26 +610,21 @@ public boolean advance() throws IOException {
           currentTimestamp = null;
+          context = new TimestampPolicyContext(false, Instant.now());
+          // queue is empty, so there is no backlog
           checkpointMark.advanceWatermark(Instant.now());
```
TODO: cannot safely advance watermark here. Use the `TimestampPolicy`

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -134,9 +135,10 @@ public RabbitMqMessage(byte[] body) {
-    this.routingKey = routingKey;
+  public RabbitMqMessage(GetResponse delivery) {
     delivery = serializableDeliveryOf(delivery);
+    this.routingKey =
```
Found what I think is a bug. The routing key should *not* come from configuration, it should come from the delivered message. A queue can be bound to multiple routing keys; it isn't a given which was used.

NJAldwin replied:
Under the "old" system, if I was listening with a routing key of `#` would this have set `routingKey` on the message to `#` (which is invalid)?  donotwant.jpg

drobert replied:
Yeah, although I think it wouldn't ultimately have any impact; I believe on publish this value is overwritten regardless. But it's still dumb.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -280,6 +282,7 @@ public int hashCode() {
         Arrays.hashCode(body),
         contentType,
         contentEncoding,
+        Objects.hashCode(headers),
```
Fixing another bug.

NJAldwin replied:
Hm, given the fact that it was left out of hashcode _and_ equals, is it possible it was intentional to not consider headers when calculating equality?

(That being said, whether or not it was intentional at the time, it makes sense to me that they should be used)

drobert replied:
> (That being said, whether or not it was intentional at the time, it makes sense to me that they should be used)

That's my opinion

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RecordIdPolicy.java:
```diff
@@ -0,0 +1,105 @@
+
+/** @todo use singletons for common policies */
+@FunctionalInterface
+public interface RecordIdPolicy extends Function<RabbitMqMessage, byte[]> {
```
The requirement that 'correlation id' is the only valid message identifier is stupid. I think messageid is more appropriate anyway, but I'm allowing for all sorts of pluggable instances.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RecordIdPolicy.java:
```diff
@@ -0,0 +1,105 @@
+    return new MessageIdPropertyPolicy();
+  }
+
+  static RecordIdPolicy body() {
```
Terrible policy really. I may remove.

NJAldwin replied:
I like that you added a warning in the javadoc in case you do leave it in.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RecordIdPolicy.java:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+
+  class MessageIdPropertyPolicy extends StringPolicy {
```
I'd actually like us to switch to this, so we can use correlation id and message id separately.

NJAldwin replied:
I am on board.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * A timestamp policy to assign event time for messages from a RabbitMq queue and watermark for it.
+ *
+ * @todo use singletons for processing time and timestamp property
```
This TODO is out of date. Remove.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,52 @@
+ * A timestamp policy to assign event time for messages from a RabbitMq queue and watermark for it.
+ *
+ * @todo use singletons for processing time and timestamp property
+ * @todo timestamp policy should include logic for determining watermark (consider accepting 'last
```
This TODO has also been implemented and can be removed.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicyFactory.java:
```diff
@@ -0,0 +1,79 @@
+   */
+  TimestampPolicy createTimestampPolicy(Optional<Instant> previousWatermark);
+
+  static TimestampPolicyFactory withProcessingTime() {
```
This is basically what we're using today, and KafkaIO has an identical one.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicyFactory.java:
```diff
@@ -0,0 +1,79 @@
+    SerializableFunction<RabbitMqMessage, Instant> timestampFunction =
+        record -> {
+          checkArgument(
+              record.getTimestamp() != null, "Rabbit message's Timestamp property is null");
```
Would be useful to be able to output *something* about the message, but :man_shrugging:

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicyFactory.java:
```diff
@@ -0,0 +1,79 @@
+        record -> {
+          checkArgument(
+              record.getTimestamp() != null, "Rabbit message's Timestamp property is null");
+          return new Instant(record.getTimestamp());
```
Note to self: we should set this property on all published messages.

Commit:
Restructuring into separate files.

CheckpointMark complete
TimestampPolicy applied to UnboundedRabbitMqReader
RecordIdPolicy appllied to UnboundedRabbitMqReader
Reworked ConnectionHandler

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMQSource.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Override
+  public boolean requiresDeduping() {
+    return true;
```
This is a change. We have an 'always unique' policy if specified, but *some* deduping strategy has to be employed in case the checkpoint mark is serialized/deserialized.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/UnboundedRabbitMqReader.java:
```diff
@@ -0,0 +1,173 @@
+
+  private final RabbitMQSource source;
+
+  private TimestampPolicy timestampPolicy;
```
Notable changes:
- contains TimestampPolicy
- contains RecordIdPolicy
- utilizes TimestampPolicyContext for tracking "last time we checked for a new message", etc.
- no longer retain currentRecordId (calculated on the fly using RecordIdPolicy)
- no longer retain currentRecordId (calculated on the fly using TimestampPolicy)
- no longer retain watermark (calculated on the fly using TimestampPolicy)
- it is acceptable to utilize the incoming CheckpointMark. It will fail at some point when it tries to finalize a checkpoint if it had previously unacked messages. If not, everything should work fine, and we have the benefit of retaining the previous watermark this way.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/UnboundedRabbitMqReader.java:
```diff
@@ -0,0 +1,173 @@
+      Channel channel = checkpointMark.getChannel();
+      checkpointMark.startReading();
+
+      // TODO: rework all of this
```
I want to get rid of 'queue vs exchange' and go all exchange+binding

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/UnboundedRabbitMqReader.java:
```diff
@@ -0,0 +1,173 @@
+        channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
+      }
+    } catch (IOException e) {
+      checkpointMark.stopReading();
```
Notable changes; lets the checkpoint mark know it can close its underlying connection if nothing is reading from it and `finalizeCheckpoint` is called

Commit:
Connection management now modeled after JDBC's

URI no longer a primitive; using ConnectionHandler supplier.
CheckpointMark now receives watermark whenever it's calculated.
Abstracted ChannelLeaser concept to use single connection per URI.

Commit:
Better throughput for rabbitmqio write

Commit:
Safer Channel management via loaner pattern

Commit:
Copious documentation

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ChannelLeaser.java:
```diff
@@ -0,0 +1,57 @@
+   * need not worry about the state of the Channel based on the thrown Exception; the {@code
+   * ChannelLeaser} implementation will be responsible for managing Connection/Channel state.
+   */
+  <T> T useChannel(UUID lesseeId, UseChannelFunction<T> f) throws IOException;
```
Note to self: settle on `lessee` vs `lesseeId`

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMQSource.java:
```diff
@@ -0,0 +1,65 @@
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+
+class RabbitMQSource extends UnboundedSource<RabbitMqMessage, RabbitMQCheckpointMark> {
```
Extracted from RabbitMqIO.java. Only notable change is `requiresDeduping`

Commit:
RecordIdPolicy tests

Commit:
Using queue depth for timestamp context

Commit:
Standardizing class names: RabbitMq*

Supporting queue depth estimates for watermark calculations

Commit:
Adding RecordId tests

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java:
```diff
@@ -0,0 +1,98 @@
+  Instant getWatermark(LastRead ctx, Instant now) {
+    if (maxEventTimestamp.isAfter(now)) {
+      return now.minus(maxDelay); // (a) above.
+    } else if (ctx.getMessageBacklog() == 0
```
Now that we have accurate (I hope?) message backlog values, this now behaves about exactly as Kafka's does

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqSource.java:
```diff
@@ -0,0 +1,65 @@
+
+  @Override
+  public boolean requiresDeduping() {
+    return true;
```
Meaningful change: this is now hard-coded to `true`

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,193 @@
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+
+class RabbitMqUnboundedReader extends UnboundedSource.UnboundedReader<RabbitMqMessage> {
```
Extracted *and* renamed *and* refactored ... sorry.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,193 @@
+      }
+
+      currentRecord = new RabbitMqMessage(delivery);
+      context = new TimestampPolicyContext(delivery.getMessageCount(), Instant.now());
```
Nifty discovery: `GetResponse` has a `messageCount` property, which is an estimate from the server of "number of messages in the queue in the Ready state". Awesome. Using this for watermark estimations.

Commit:
Minor same instance check to CheckpointMark

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,193 @@
+          source.spec.timestampPolicyFactory().createTimestampPolicy(prevWatermark);
+    }
+    return this.timestampPolicy;
+  }
```
⛏ this method sets `this.timestampPolicy` directly, but it also returns the value, which is unused.  imo it could be clearer as something like
```java
    ensureTimestampPolicy(
        Optional.ofNullable(checkpointMark).map(RabbitMqCheckpointMark::getWatermark));
  }

  private void ensureTimestampPolicy(Optional<Instant> prevWatermark) {
    if (this.timestampPolicy == null) {
      this.timestampPolicy =
          source.spec.timestampPolicyFactory().createTimestampPolicy(prevWatermark);
    }
  }
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,193 @@
+      // we consume message without autoAck (we want to do the ack ourselves)
+      GetResponse delivery =
+          connectionHandler.useChannel(
+              checkpointMark.getCheckpointId(), (channel) -> channel.basicGet(queueName, false));
```
I know this already existed, but while you're here?  💅 
```suggestion
              checkpointMark.getCheckpointId(), (channel) -> channel.basicGet(queueName, /* autoAck: */ false));
```

NJAldwin replied:
also, a nit: the parens around `(channel)` should be unnecessary

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * A timestamp policy to assign event time for messages from a RabbitMq queue and watermark for it.
+ *
+ * <p>Inspired by a similar concept if KafkaIO
```
⛏ 
```suggestion
 * <p>Inspired by a similar concept in KafkaIO
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public abstract class TimestampPolicy {
+
+  /** Analgous to KafkaIO's TimestampPolicy.PartitionContext */
```
⛏ 
```suggestion
  /** Analagous to KafkaIO's TimestampPolicy.PartitionContext */
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+      try {
+        toClose.close();
+      } catch (IOException | TimeoutException e) {
+        // ignore
```
is there any case wherein `close()` would throw an exception without actually closing the channel?  the docs don't seem to say anything about it, so may be not worth worrying about, but just wondering since we're swallowing the exception here

drobert replied:
On the off chance it did, it would ultimately be closed as part of closing the Connection after this. But almost assuredly if this fails, the connection/channel have already been severed.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+          connectionFactory.setRequestedChannelMax(0);
+          connectionFactory.setRequestedFrameMax(0);
+        } catch (URISyntaxException e) {
+          // full URI excluded lest it contain user/pass
```
👍 
wonder if it would be useful to add `(URI redacted)` to the exception

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+
+    return channelsByLessee.computeIfAbsent(
+        lessee,
+        (uuid) -> {
```
So it doesn't become tiresome, I'm not gonna comment on any more cases of unnecessary parens for single-param lambdas :)

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+          try {
+            return connection
+                .openChannel()
+                .orElseThrow(() -> new RuntimeException("No RabitMQ channel available"));
```
```suggestion
                .orElseThrow(() -> new RuntimeException("No RabbitMQ channel available"));
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+                .openChannel()
+                .orElseThrow(() -> new RuntimeException("No RabitMQ channel available"));
+          } catch (IOException e) {
+            throw new RuntimeException("No RabitMQ channel available");
```
```suggestion
            throw new RuntimeException("No RabbitMQ channel available");
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+   *     while closing individual Channels are ignored
+   */
+  @Override
+  public synchronized void close() throws IOException {
```
`synchronized` 👍

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,157 @@
+
+  private Channel getChannel(UUID lessee) throws IOException {
+    if (connection == null) {
+      synchronized (this) {
```
I'm a little rusty on my concurrency principles, but I think you need another `if (connection == null)` inside this `synchronized` in case:

A: calls `getChannel` -> sees `connection == null`
B: calls `getChannel` -> sees `connection == null`
A: enters synchronized block, builds connection, leaves synchronized block
B: enters synchronized block, builds new connection, leaves synchronized block

NJAldwin replied:
alternatively, could make `connection` an `AtomicReference`

drobert replied:
making it an AtomicReference I *think* would still allow for me to create a new Channel that may potentially go unused. I do think you're correct about this requiring an additional `if null` check, however.

NJAldwin replied:
I was thinking `AtomicReference` had something on it that would allow for this sort of behavior, but I was mistaken.  I would actually probably use a memoized `Supplier` (w/ Guava's `Suppliers#memoize`) save for the fact that this needs to be null-tolerant for serialization/deser

drobert replied:
Yeah, if AtomicReference had a *functional* lazy creator it would work, but it doesn't...

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqCheckpointMark.java:
```diff
@@ -0,0 +1,149 @@
+          for (Long deliveryTag : deliveryTags) {
+            channel.basicAck(deliveryTag, false);
+          }
+          deliveryTags.clear();
```
I don't know if we expect `finalizeCheckpoint` to at all potentially be called multiple times, but if we did, we probably would want to wrap the list in `synchronizedList` to make concurrent access safe.  e.g. here we'd want to extract all the values and clear the list before starting to iterate over the values

drobert replied:
I don't believe there is any risk of a checkpoint mark being used concurrently like that. It should be assigned to at most one reader at any given time.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqCheckpointMark.java:
```diff
@@ -0,0 +1,149 @@
+  /**
+   * When a CheckpointMark is serialized, deserialized, and passed into the constructor of a new
+   * Reader, it's imperative the reader and the checkpoint mark utilize the same channel. This is
+   * called by the Reader to enforce this.
```
⛏ `This method is called by the Reader to enforce this constraint`

disambiguating the two different `this`s in this sentence

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -28,11 +28,14 @@
+ * Serializable wrapper around rabbitmq's {@link GetResponse}.
+ *
+ * <p>The main purpose of this class is that AMQP.BasicProperties doesn't provide a serializable
+ * public API.
```
Super nitpick:

> The main reason for this class is that that AMQP.BasicProperties doesn't provide a serializable public API

or

> The main purpose of this class is to provide a serializable public API for AMQP.BasicProperties, which lacks one of its own.

drobert replied:
This was pre-existing, but I'll adopt a clarifying change. I have much documentation to write/update still.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqSource.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public List<RabbitMqSource> split(int desiredNumSplits, PipelineOptions options) {
+    // RabbitMQ uses queue, so, we can have several concurrent consumers as source
+    List<RabbitMqSource> sources = new ArrayList<>();
+    for (int i = 0; i < desiredNumSplits; i++) {
+      sources.add(this);
+    }
+    return sources;
```
Could replace this with a stream.  something like
```java
    return Stream.generate(() -> this)
        .limit(desiredNumSplits)
        .collect(Collectors.toList());
```

drobert replied:
Could. Nothing else in Beam seems to be written that way, however, and I'm not seeing much benefit.

NJAldwin replied:
does beam predate java streams? :)

imo it's clearer, too, but ¯\\_(ツ)_/¯

drobert replied:
It does predate it. For something this simple I don't have a strong opinion. I mostly want to avoid any stylistic conversations with the Beam folks since this is such a significant behavioral rewrite.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RecordIdPolicy.java:
```diff
@@ -0,0 +1,122 @@
+public interface RecordIdPolicy extends Function<RabbitMqMessage, byte[]> {
+
+  /** @return a policy that defines the message id by the amqp {@code correlation-id} property */
+  static RecordIdPolicy correlationId() {
```
man, Scala's companion objects have spoiled me.  At first glance I missed `static` and was like "how can this `@FunctionalInterface` have more than one method"

drobert replied:
Yeah, I wanted to make these all singletons and I kept forgetting or not caring how best to do that in java. This whole thing was a big refresher course for me.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicyFactory.java:
```diff
@@ -0,0 +1,79 @@
+  static TimestampPolicyFactory withTimestampProperty(Duration maxDelay) {
+    SerializableFunction<RabbitMqMessage, Instant> timestampFunction =
+        record -> {
+          checkArgument(
```
there is a `checkNotNull` but I'm guessing you're using this so it's an `IAE` rather than `NPE`?

drobert replied:
Mostly was following suit with other implementations. Ultimately I don't think it matters what exception is thrown.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/package-info.java:
```diff
@@ -16,5 +16,5 @@
  */
 
-/** Transforms for reading and writing from RabbitMQ. */
+/** Transforms for reading and writing from RabbitMQ (AMQP 0.9). */
```
👍

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public class ConnectionProviderFromUri
+    implements SerializableFunction<Void, ConnectionHandler>, HasDisplayData, Closeable {
+  private static final ConcurrentHashMap<String, ConnectionHandler> instances =
```
totally fine if this isn't the "beam" style, but I know in the Google style, `psf`s should be treated as constants for naming: `INSTANCES`

drobert replied:
I was following suit from the JDBC provider, which used this naming convention.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java:
```diff
@@ -0,0 +1,102 @@
+    this.uri = uri;
+    String displayable = uri;
+    try {
+      URI parsed = URI.create(uri);
```
`URI.create` throws an IAE if the given URI violates the RFC.  This should probably be `new URI(uri)`.

```suggestion
      URI parsed = new URI(uri);
```

Alternatively, we could separately catch an IAE, but URI#create literally just wraps the URISyntaxException:
```java
    public static URI create(String str) {
        try {
            return new URI(str);
        } catch (URISyntaxException x) {
            throw new IllegalArgumentException(x.getMessage(), x);
        }
    }
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java:
```diff
@@ -0,0 +1,102 @@
+          Optional.ofNullable(parsed.getUserInfo())
+              .map(x -> x.split(":", 2)[0]) // user:pass
+              .map(username -> username + ":")
+              .orElse("");
```
⛏ 
took me a sec to realize this is stripping the password.  might be straightforward to extract it to a function:
```java
      String userInfo =
          Optional.ofNullable(parsed.getUserInfo())
              .map(this::stripPassword)
              .orElse("");

...


  private String stripPassword(String userInfo) {
    return userInfo.split(":", 2)[0] + ":";
  }
```

drobert replied:
Yeah, this was surprisingly difficult to figure out how to do.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @Override
+  public void close() throws IOException {
```
IJ tells me you can omit this `throws IOException`

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java:
```diff
@@ -0,0 +1,98 @@
+/**
+ * A policy for custom record timestamps where timestamps within the queue are expected to be
+ * roughly monotonically increasing with a cap on out of order event delays (say 1 minute). The
+ * watermark at any time is '({@code Min(now(), Max(event timestamp so far)) - max delay})'.
```
```suggestion
 * watermark at any time is '({@code earliest(now(), latest(event timestamp so far)) - max delay})'.
```

I like to use "earliest" and "latest" rather than "min" and "max" when comparing time

(whereas "max delay" is a param so that stays)

NJAldwin replied:
(there's javadoc below with a similar pseudocode)

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java:
```diff
@@ -0,0 +1,98 @@
+ * watermark at any time is '({@code Min(now(), Max(event timestamp so far)) - max delay})'.
+ * However, watermark is never set in future and capped to 'now - max delay'. In addition, watermark
+ * advanced to 'now - max delay' when the queue has caught up (previous read attempt returned no
+ * message)
```
⛏ this sentence seems to be missing some articles

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java:
```diff
@@ -0,0 +1,98 @@
+   * is {@code Min(now(), max_event_timestamp) - maxDelay}.
+   *
+   * @param timestampFunction A function to extract timestamp from the record
+   * @param maxDelay For any record in the Kafka partition, the timestamp of any subsequent record
```
I see this was copied from `Kafka` :) but we should probably change this to "Rabbit queue"
```suggestion
   * @param maxDelay For any record in the Rabbit queue, the timestamp of any subsequent record
```

drobert replied:
ha! thanks

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -230,18 +170,45 @@ public void stop() throws IOException {
+     *     required.
+     */
     public Read withUri(String uri) {
       checkArgument(uri != null, "uri can not be null");
```
should we add another check here to make sure that the URI parses?  or do we not care since we're just passing it down anyways (with the assumption that there could possibly be URIs that violate the RFC but the rabbit client still accepts)

drobert replied:
The general Beam approach seems to be to let it fail at runtime. I think in the case of this Builder pattern it's best validated at either build() time or execution time.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -345,240 +312,56 @@ public Read withMaxReadTime(Duration maxReadTime) {
+     * Sets the timestamps policy based on the amqp basic "timestamp" property of the message. It is
+     * an error if a record's timestamp property is not populated. The timestamps within a queue are
+     * expected to be roughly monotonically increasing with a cap on out of order delays (e.g. 'max
+     * delay' of 1 minute). The watermark at any time is '({@code Min(now(), Max(event timestamp so
```
same comment about Min/Max -> earliest/latest from above (feel free to ignore)

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -345,240 +312,56 @@ public Read withMaxReadTime(Duration maxReadTime) {
-      }
-      return advance();
-    }
+      PTransform<PBegin, PCollection<RabbitMqMessage>> transform = unbounded;
```
imo this would be clearer as 
```java
PTransform<PBegin, PCollection<RabbitMqMessage>> transform = (maxNumRecords() < Long.MAX_VALUE || maxReadTime() != null)
  ? unbounded.withMaxReadTime(maxReadTime()).withMaxNumRecords(maxNumRecords())
  : unbounded;
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -699,49 +504,73 @@ public Write withQueueDeclare(boolean queueDeclare) {
-        if (spec.queue() != null && spec.queueDeclare()) {
-          channel.queueDeclare(spec.queue(), true, false, false, null);
+        if (channelLeaser == null) {
+          channelLeaser = spec.connectionHandlerProviderFn().apply(null);
```
man, annoying beam doesn't have a `SerializableSupplier` or `SerializableProvider`.  passing `null` here is slightly misleading

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -699,49 +504,73 @@ public Write withQueueDeclare(boolean queueDeclare) {
+                channel.exchangeDeclare(spec.exchange(), spec.exchangeType());
+              }
+              if (spec.queue() != null && spec.queueDeclare()) {
+                channel.queueDeclare(spec.queue(), true, false, false, null);
```
can you pop a comment on here similar to the comment in the reader with the param names for these booleans? (someone didn't read The Boolean Trap before writing this api :( )

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -699,49 +504,73 @@ public Write withQueueDeclare(boolean queueDeclare) {
+
+          ChannelLeaser.UseChannelFunction<Void> basicPublishFn =
+              (channel) -> {
+                // TODO: what's with this ridiculous hard-coding of message props?
```
![kill-it-with-fire](https://user-images.githubusercontent.com/145036/71425362-d37fa180-2669-11ea-90e8-ef5a5d2130d5.gif)

NJAldwin reviewed (changes requested):
I love this.

One serious concurrency concern (could leak connections if I'm reading it right) and one concurrency question.

I think most of the rest of my comments are nitpicks / java style / etc.

Commit:
PR feedback, mostly nit-picks

Commit:
Mostly documentation updates

Commit:
Simplifying Write

Commit:
Converting RabbitMqMessage to AutoValue/Builder

Removing exchange declare from RabbitMqIO.
Removing publishing to 'queue' concept in RabbitMqIO.

Commit:
Tests compiling

Commit:
Compiles and tests run.

Added totalBacklogBytes support to rabbitmq io.
Cleaned up getWatermark() public api.
Removed legacy record id policies.
Updated a few record id policy tests.

Commit:
Fixing some broken record id tests

Commit:
Direct exchange test is passing.

RabbitMq Read/Write factories now require a uri to even begin.
Tests now more oriented around AmqpMessage rather than payloads.

Commit:
Reworking in terms of ReadParadigm to manage invariants

Commit:
Cleanup

drobert commented on sdks/java/io/rabbitmq/build.gradle:
```diff
@@ -31,6 +31,8 @@ dependencies {
   testCompile "org.apache.qpid:qpid-broker-core:7.1.5"
   testCompile "org.apache.qpid:qpid-broker-plugins-memory-store:7.1.5"
   testCompile "org.apache.qpid:qpid-broker-plugins-amqp-0-8-protocol:7.1.5"
+  testCompile library.java.hamcrest_core
```
Other IOs have these. They make for nicer assertions in unit tests (`assertThat(foo, isLike(some other thing))`

NJAldwin replied:
💯 hamcrest and assertj make testing in java so much nicer

Commit:
Documentation examples updates

Commit:
Removing write(uri)

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -53,276 +38,235 @@
   /** A {@link PTransform} to consume messages from RabbitMQ server. */
   @AutoValue
   public abstract static class Read extends PTransform<PBegin, PCollection<RabbitMqMessage>> {
-    @Nullable
```
Instead of a bunch of disparate properties to aggregate together somehow later, I've broken these into a known set of `ReadParadigm`s. ... my kingdom for a `sealed trait` here.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -17,24 +17,168 @@
+ * with AutoValue makes this a bit easier to work with.
  */
-public class RabbitMqMessage implements Serializable {
+@AutoValue
```
I've rebuilt this using google's AutoValue library which is used liberally throughout this code base. I wanted to be able to modify individual properties (like routing key) easily.

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+
+      ChannelLeaser.UseChannelFunction<String> setupFn =
+          channel -> {
+            ReadParadigm paradigm = source.spec.readParadigm();
```
Uses the new `ReadParadigm`

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+  }
+
+  @Override
+  public long getTotalBacklogBytes() {
```
Look, backlog estimation!

NJAldwin replied:
worth putting a bit of javadoc on this to state that this is just an estimation?

drobert replied:
I suppose

drobert commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,93 @@
+   * plugin</a>, which first looks for header {@code timestamp_in_ms} then falls back to amqp
+   * message property {@code timestamp}
+   */
+  public static SerializableFunction<RabbitMqMessage, Instant>
```
Extracted this extremely useful, common instance of extracting a timestamp. Seems useful.

drobert commented on sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/ExchangeTestPlan.java:
```diff
@@ -24,11 +24,7 @@
- * queue name match.
- *
- * <p>To facilitate the many combinations of queue bindings, routing keys, and exchange declarations
+ * To facilitate the many combinations of queue bindings, routing keys, and exchange declarations
```
TODO: totally rework 'exchange testing'

drobert commented on sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIOTest.java:
```diff
@@ -98,60 +98,65 @@ public static void beforeClass() throws Exception {
-  @Test
-  public void testReadQueue() throws Exception {
+  @Test(timeout = ONE_MINUTE_MS)
+  public void testReadDefaultExchangeExistingQueue() throws Exception {
```
This is the only test I've as-of-yet rewritten and have seen it pass just to prove basic functionality.

drobert commented on sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqTestUtils.java:
```diff
@@ -82,9 +183,8 @@ public TestConsumer(Channel channel, List<String> received) {
-        String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body)
-        throws IOException {
-      received.add(recordToString(body));
+        String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) {
```
Hmm, these should be full `RabbitMqMessage` instances.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -53,276 +38,235 @@
+ * non-issue).
  *
- * }</pre>
+ * <p>Because Beam runners are free to close readers and start new ones, it's *highly* likely,
+ * especially in an unbounded streaming reader, that messages will be re-delivered (e.g. channel was
+ * closed before the messages were acknowledged and were re-read on a new reader). For this reason,
+ * it's *highly* encouraged callers specify an appropriate {@link RecordIdPolicy}. The default is to
+ * allow for messages to be replayed, which may be acceptable if your pipeline is built for
+ * at-least-once data processing, but is likely undesirable. The safest approach would be to utilize
+ * the amqp property {@code messageId}, ensure it's set to a globally unique value such as a UUID,
+ * and use {@link RecordIdPolicy#messageId()}.
```
:+1:

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -53,276 +38,235 @@
+        new URI(uri);
+      } catch (URISyntaxException e) {
+        throw new IllegalArgumentException(
+            "Malformed rabbit mq supplied (omitted here lest it contain username/password)");
```
⛏ 
```suggestion
            "Malformed rabbit mq uri supplied (omitted here lest it contain username/password)");
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -345,240 +289,92 @@ public Read withMaxReadTime(Duration maxReadTime) {
+    /**
+     * Sets {@link TimestampPolicy} to {@link TimestampPolicyFactory.ProcessingTimePolicy}. This is
+     * the default timestamp policy. It assigns processing time to each record. Specifically, this
+     * is the timestamp when the record becomes 'current' in the reader. The watermark aways
```
⛏ 
```suggestion
     * is the timestamp when the record becomes 'current' in the reader. The watermark always
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -345,240 +289,92 @@ public Read withMaxReadTime(Duration maxReadTime) {
+     *
+     * <p>However, the watermark is never set to a timestamp in the future and is capped to 'now -
+     * max delay'. In addition, the watermark * is advanced to 'now - max delay' when the queue has
+     * caught up (previous read attempt returned no * message and/or estimated backlog per {@code
```
⛏  here & line above: some errant `*`s that snuck in

NJAldwin replied:
also in the same place for `withTimestampPluginCompatPolicyFactory`'s javadoc

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -345,240 +289,92 @@ public Read withMaxReadTime(Duration maxReadTime) {
-      }
-      return currentTimestamp;
+    /**
+     * Sets a timestamp policy based on a custom timestamp extration strategy where the timestamps
```
⛏ 
```suggestion
     * Sets a timestamp policy based on a custom timestamp extraction strategy where the timestamps
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java:
```diff
@@ -654,94 +434,59 @@ public Write withExchange(String exchange, String exchangeType) {
+          channelLeaser = spec.connectionHandlerProviderFn().apply(null);
         }
+
+        channelLeaser.useChannel(writerId, ch -> null);
```
is this here just so you can make sure a connection has been made?  might be worth a comment as it looks a little strange at first glance

NJAldwin replied:
also tiny suggestion -- Java lacks `_` but I've mostly seen `ignored` used instead:
```suggestion
        channelLeaser.useChannel(writerId, ignored -> null);
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -17,24 +17,168 @@
+  public static Builder builder() {
+    return new AutoValue_RabbitMqMessage.Builder()
+        .setDeliveryMode(1) // non-durable
+        .setPriority(1) // lowest priority
```
minor, but for these two can we make constants so this becomes
```java
.setDeliveryMode(NON_DURABLE_DELIVERY)
.setPriority(LOWEST_PRIORITY)
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -17,24 +17,168 @@
+        .setPriority(1) // lowest priority
+        .setRoutingKey("")
+        .setTimestamp(
+            new Date(java.time.Instant.now().truncatedTo(ChronoUnit.SECONDS).toEpochMilli()))
```
this might warrant a comment, something along the lines of `// rabbit natively only supports second-granularity timestamps`

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -17,24 +17,168 @@
+    delivery = serializableDeliveryOf(delivery);
+    BasicProperties props = delivery.getProps();
+    Map<String, Object> headers = new HashMap<>();
+    if (headers == null) {
```
Am I missing something here?  `headers` should always be non-null (as it's set on the line above), so L107-110 will never run, right?

drobert replied:
I must have lost something in a refactor. I'll take another pass. But yeah, this certainly appears to be impossible :-)

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -17,24 +17,168 @@
+      Date timestamp = timestamp();
+      if (timestamp != null) {
+        setTimestamp(
+            new Date(timestamp.toInstant().truncatedTo(ChronoUnit.SECONDS).toEpochMilli()));
```
ah I see this is repeated -- might be worth extracting to a util fn (and the comment mentioned above can go on that fn)

drobert replied:
I don't actually need both now that i've modified the `build()` operation to perform normalization for me. I'll just call `setTimestamp(new Date())` here and add the comments to the normalization in `build()`

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+              if (!"".equalsIgnoreCase(newQueue.getExchange())) {
+                channel.queueBind(queueName, newQueue.getExchange(), newQueue.getRoutingKey());
+              }
+            }
```
not sure it's worth the refactor, but in general when building something like this in Java, rather than `instanceof` & casting, I'd actually put this behavior on a method on `ReadParadigm` -- something like `paradigm.bindAsNecessary(channel)`

drobert replied:
I like it. I really wanted a sealed trait or an enum, but no can do. I'll take a whack at it because it did feel pretty dirty.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+              // this may fail if the queue isn't actually new and has different properties
+              channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+
+              // the default exchange does not allow for explicitly binding a queue
```
fun gotcha

drobert replied:
Yeah, fun new discovery for me during testing.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+
+      queueName = connectionHandler.useChannel(checkpointMark.getCheckpointId(), setupFn);
+    } catch (IOException e) {
+      checkpointMark.stopReading();
```
it's effectively the same right now, but I wonder if it would be clearer to call `close()` in these catch blocks

drobert replied:
Maybe, but I don't think so. I don't know all the intended semantics of `close` for a Reader in Beam, but I do know that I want to inform the checkpoint mark that nothing is attempting to read from it. It's possible Beam could call `start` again in which case this would definitely not be "closed".

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java:
```diff
@@ -0,0 +1,230 @@
+      throw e;
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
```
`start()` calls `checkpointMark.stopReading()` on exceptions, but `advance()` here does not.  is that intentional?

drobert replied:
Yes. If we fail to read on start I think the reader can be considered unusable. in 'advance', it seems reasonable that the Beam runner could re-try advancing, in which case we really only want to tell the checkpoint mark we're done reading when the runner calls 'close()'

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/TimestampPolicy.java:
```diff
@@ -0,0 +1,93 @@
+            Date timestamp = record.timestamp();
+            if (timestamp == null) {
+              throw new IllegalArgumentException(
+                  "Neither timestamp_in_ms header nor timestamp property contain a valid timestamp value");
```
the javadoc on `RabbitMqIO` stated that this would error if both were not present.  looks like it actually only errors if _neither_ is present?

drobert replied:
That is correct. Will update comments and code accordingly.

drobert replied:
I'm seeing this in `RabbitMqIO`:

> It is an error if a record's {@code timestamp} property is not
>     * populated and the {@code timestamp_in_ms} header is missing or malformed

That sounds accurate to me but I suppose I can see how this is misleading. Suggestions on how to phrase this?

NJAldwin replied:
Looking back at that, you're right that it's not saying both are required -- I must have misread.

Perhaps something like
> At least one of the `timestamp` property and the `timestamp_in_ms` header must be present and well-formed for a record to be processed.

¯\\_(ツ)_/¯

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ReadParadigm.java:
```diff
@@ -0,0 +1,224 @@
+ * <p>Some known limitations of these approaches:
+ *
+ * <ul>
+ *   <li>RabbitMqIO intentionally does not support exchange declaration. This feels out of scope and
+ *       there's unlikely a good use case for having beam declare the exchange along with the queue.
```
👍

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ReadParadigm.java:
```diff
@@ -0,0 +1,224 @@
+
+    public static ExistingQueue of(String queueName) {
+      checkArgument(
+          queueName != null && !queueName.trim().isEmpty(),
```
seeing as how Guava is available here, this could be `!Strings.isNullOrEmpty(queueName)`

drobert replied:
Hmm, I thought that was the case, but diving in, I don't see a "trim" operation. I think that's fine, but just noting.

drobert replied:
Ah, I forgot about that. Spring had a StringUtils that would perform trim but guava's didn't. I actually think I'm going to leave it.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ReadParadigm.java:
```diff
@@ -0,0 +1,224 @@
+   */
+  public static class NewQueue extends ReadParadigm {
+    private final String exchange;
+    private final String exchangeType;
```
I don't feel _especially_ strongly about this but it might be nice for this to be an enum

drobert replied:
I think I can just use the underlying lib's data type for that.

drobert replied:
Better: I don't actually use this property anywhere. Now that I've specified reading happens from a queue, I only need to know the type to validate inputs, and I have specialized factory functions for that. Excellent.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java:
```diff
@@ -0,0 +1,110 @@
+
+  public ConnectionProviderFromUri(String uri) {
+    this.uri = uri;
+    String displayable = uri;
```
if the given URI contains sensitive user info, but is unparseable as a URI, this has the potential to leak the credentials, right?

drobert replied:
yes, it's possible. I think what I have is better than what exists, but I'm not sure how much sleep I want to lose over this.

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,179 @@
+                .openChannel()
+                .orElseThrow(() -> new RuntimeException("No RabbitMQ channel available"));
+          } catch (IOException e) {
+            throw new RuntimeException("No RabbitMQ channel available");
```
```suggestion
            throw new RuntimeException("No RabbitMQ channel available", e);
```

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java:
```diff
@@ -0,0 +1,179 @@
+    } catch (RuntimeException e) {
+      // may have been wrapped
+      Throwable cause = e.getCause();
+      if (cause instanceof ShutdownSignalException) {
```
I dunno how deeply this may have been wrapped (so my guess is a single level is fine) but fyi there does exist a way to search the whole chain:
```java
Option<ShutdownSignalException> sse = Throwables.getCausalChain(e).stream().filter(t -> t instanceof ShutdownSignalException).findFirst().map(t -> (ShutdownSignalException)t);
```

Commit:
Minor test refactoring

Commit:
ExchangeType now an enum.

Typos/PR feedback

Commit:
Removing obsolete queue name check

Commit:
Checkstyle

Commit:
Removing unused test functions

Commit:
RabbitMqMessage public

Commit:
RabbitMqMessage contains equals/hashCode

Upgrading some deps

Commit:
Remove any pretense of re-using a RabbitMqCheckpointMark

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java:
```diff
@@ -42,6 +44,7 @@
  * with AutoValue makes this a bit easier to work with.
  */
 @AutoValue
+@DefaultCoder(SerializableCoder.class)
```
👀

NJAldwin commented on sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqSource.java:
```diff
@@ -50,7 +50,11 @@
   public UnboundedReader<RabbitMqMessage> createReader(
       PipelineOptions options, RabbitMqCheckpointMark checkpointMark) throws IOException {
-    return new RabbitMqUnboundedReader(this, checkpointMark);
+    // per conversation in Beam Dev DL, an 'ack'-style CheckpointMark cannot safely
+    // be re-used so can and should be ignored during construction
+    // see
+    // http://mail-archives.apache.org/mod_mbox/beam-dev/201911.mbox/%3cCAFmTo4_cw6xqLw+8VpJEL0fwFY3M4HS173KKmQLs-KPAkh+Bvg@mail.gmail.com%3e
+    return new RabbitMqUnboundedReader(this);
```
👍

drobert commented:
Closing this. I'm ready to submit this upstream for much feedback.
